test/controllers/predicates/SaveOriginAndRedirectSpec.scala

```diff
@@ -0,0 +1,97 @@
+import play.twirl.api.Html
+import testConstants.BaseTestConstants.{testListLink, testMtditid, testNino, testRetrievedUserName}
+import testConstants.IncomeSourceDetailsTestConstants.singleBusinessIncome
+import testUtils.TestSupport
+import uk.gov.hmrc.http.SessionKeys
+import views.html.navBar.BtaNavBar
+
+import scala.concurrent.Future
+
+class SaveOriginAndRedirectSpec extends TestSupport with MockAsyncCacheApi with FeatureSwitching {
+
+  lazy val successResponseWithoutOrigin: MtdItUser[Any] = MtdItUser(testMtditid, testNino, Some(testRetrievedUserName), singleBusinessIncome,
+    None, Some("testUtr"), Some("testCredId"), Some("Individual"), None)
+
+  lazy val successResponseWithBtaOriginAndWithoutSession: MtdItUser[Any] = MtdItUser(testMtditid, testNino, Some(testRetrievedUserName), singleBusinessIncome,
+    Some(testView.apply(testListLink)), Some("testUtr"), Some("testCredId"), Some("Individual"), None)(fakeRequestQueryStringAndOriginWithoutSession("BTA"))
+
+  lazy val successResponseWithSessionOriginPTA: MtdItUser[Any] = MtdItUser(testMtditid, testNino, Some(testRetrievedUserName), singleBusinessIncome,
+    Some(Html("")), Some("testUtr"), Some("testCredId"), Some("Individual"), None)(fakeRequestQueryStringAndOrigin("BTA", "PTA"))
+  val testView: BtaNavBar = app.injector.instanceOf[BtaNavBar]
+
+  lazy val successResponseWithInvalidQueryString: MtdItUser[Any] = MtdItUser(testMtditid, testNino, Some(testRetrievedUserName), singleBusinessIncome,
+      None, Some("testUtr"), Some("testCredId"), Some("Individual"), None)(fakeRequestQueryStringAndOriginWithoutSession("INVALID"))
+
+  val obj: SaveOriginAndRedirect = new SaveOriginAndRedirect() {
+    override def messagesApi: MessagesApi = mock[MessagesApi]
+
+    override val appConfig: FrontendAppConfig = mock[FrontendAppConfig]
+  }
+
+  def fakeRequestQueryStringAndOrigin(origin: String, session: String): FakeRequest[AnyContentAsEmpty.type] = fakeRequestWithQueryString(origin).withSession(
+    "origin" -> session)
+
+  def fakeRequestQueryStringAndOriginWithoutSession(origin: String): FakeRequest[AnyContentAsEmpty.type] = fakeRequestWithQueryString(origin)
+
+  def fakeRequestWithQueryString(origin: String): FakeRequest[AnyContentAsEmpty.type] = FakeRequest(GET, s"test/url?origin=$origin")
+
+  "The SaveOriginAndRedirect" when {
+
+    "Valid request is passed" should {
+      "redirect to call with session origin changes to BTA" in {
+        val result: Future[Result] = obj.saveOriginAndReturnToHomeWithoutQueryParams(successResponseWithSessionOriginPTA, false)
+        result.futureValue.header.status shouldBe 303
+        result.futureValue.session.get("origin") shouldBe Some("BTA")
+      }
+      "redirect to call with adding origin BTA to session" in {
+        val result: Future[Result] = obj.saveOriginAndReturnToHomeWithoutQueryParams(successResponseWithBtaOriginAndWithoutSession, false)
+        result.futureValue.header.status shouldBe 303
```

Comment:
minor, if you can replace => 303 with SEE_OTHER from 
play.api.http.Status, as we should avoid using "magic numbers"